RFC - Graph Type System (#352)
blockprotocol/blockprotocol: rfcs/text/0352-graph-type-system.md

  "http://example.com/property-types/age": 30,
  "http://example.com/property-types/planetOfOrigin": "Earth",
  "http://example.com/property-types/occupation": "Intergalactic Traveler"
}
```

> 💭 Acknowledgement: This is less ergonomic than the current state of things; this drawback among others, is explored with some possible mitigations in the [Drawbacks](#drawbacks), [Rationale and Alternatives](#rationale-and-alternatives), and [Future Possibilities](future-possibilities) sections

This will have impact on how developers:

- **access** fields in entities
- format properties when they **create** entities
- format properties when they **update** entities

### Interfacing with Linked Entities

#### Receiving Links and Linked Entities

Links will likely continue to be returned in a separate collection alongside entities when blocks receive data. The data within the `linkGroups` object might change to incorporate the new type information, likely with the `path` field being replaced with something that uses the link type's URI.

#### Creating Links

The Link Functions will be updated to handle Link Types, removing the `path` fields and instead using the URI of the Link Type.

> ### block Schemas

Author: ```suggestion
### Block Schemas
```

- Feature Name: graph-type-system
- Start Date: 2022-06-14
- RFC PR: [blockprotocol/blockprotocol#352](https://github.com/blockprotocol/blockprotocol/pull/352)
- Block Protocol Discussion: [blockprotocol/blockprotocol#0000](https://github.com/blockprotocol/blockprotocol/discussions/0000)

# Summary

[summary]: #summary

This RFC proposes changing the way data structural requirements are expressed within the BP.

> Rather than allowing any free-form JSON schema, this RFC defines a set of JSON schema meta-schemas, which describe:

Reviewer: Cool, have made the change in https://github.com/blockprotocol/blockprotocol/pull/352/commits/736eeb99639134af39bd5fe12dc4f74fc32f3d97

Have slightly tweaked the added sentence further:
> The structure of these types are defined by (and can be validated against) their meta-schemas, which can be found in the [reference-level explanation](#reference-level-explanation).

My thinking behind referring to them as "meta-schemas" rather than "JSON schema meta-schemas" is because I find its too similar to "JSON schema meta-schema", which I would interpret as a reference to the meta-schema of JSON schema (draft-07, draft-05, etc). This also feels like it better matches how "meta-schemas" are referred to in the [JSON-Schema spec](https://json-schema.org/specification.html).

> Rather than allowing any free-form JSON schema, this RFC defines a set of JSON schema meta-schemas, which describe:

Author: That's fair enough, I'm fine with the change, although I think meta-schema on it's own is pretty vague and we're specifically talking about a JSON Schema which _is_ a meta-schema (i.e. it validates other JSON schema)

Anyway changes are good

By getting both parties to agree to a prescriptive standard, new technological possibilities become viable. The Block Protocol is an example of this, enabling a new class of applications, ones that are able to dynamically load complex front-end components at run-time which are immediately usable and functional, even if the component was previously unknown to the application.

> The foundation of this capability is the standardization of the expression of requirements. Most notably, the block and entity schemas expressed through the graph service allow for dynamic resolution of expected input data.

Reviewer: Added in https://github.com/blockprotocol/blockprotocol/pull/352/commits/8fc2b667887c6e0a57fd5b8e67f60b71de18b826

---

**Example**:

A block expresses that it needs a URL hosting an image (presumably so that it can render it):

```json
{
  "properties": {
    "image_url": "string"
  }
}
```

The embedding application is then able to search across its data to find entities that satisfy this requirement, such as:

```json
{
  "image_url": "https://dummyimage.com/300.jpg",
  ... // some other fields

> }

Author: @thehabbos007 and I were discussing this as well, we're not sure what the best solution is 🙃 
We tried `jsonc` but it actually removes a lot of the formatting. 

The big red boxes have grown on me if I'm honest, they kind of make it very clear that it's not part of the data, but I appreciate they could be very distracting. 

Are you able to attach a screenshot here showing the difference between JSON and JS formatting?

> }

Reviewer: I don't feel super strongly about this, but I do find the red boxes distracting. And I think it's better for the reader to be focusing on the actual code in the document, rather than our comments and `...` taking the centre stage.

But again it's just a preference.

> }

Reviewer: Fair enough, don't feel strongly so will resolve.

- a **required** definition of its possible value

#### Primitive Top-Level Data Types

We pre-define a set of primitive top-level Data Types.
These can be considered the least restrictive value spaces:

- `Text`
- `Number`
- `Boolean`
- `Null`
- `Object`
- `Empty List`

#### Creating new Data-Types

Due to the size of this RFC, creating new data-types has been deemed out-of-scope and a [follow-up RFC](https://github.com/blockprotocol/blockprotocol/pull/355) is in the works to spec out how this could work. Please consult it for plans on introducing a suite of Data Types including ones like `Date`, `Positive Number`, etc.

### Property Types

A **Property Type** is a description of a named piece of data, including its possible values. A Property Type is composed of the following:

- a **required** `$id`, which is a globally unique identifier, where the property type's definition can be accessed (in most cases this will be a URL)
- a **required** `name`, which should generally be a non-pluralized description of the property (for example: "Address" not "Addresses")
- **optionally** a `description` to further explain the semantic meaning of the property type

> - a **required** definition of its possible values

Author: Not an accidental copy. The "required definition of its possible values" is the rest of the section.

i.e.
> **Defining the possible values of a property**

**Defining the possible values of a property**

#### Data Types

A Property Type can have plain data as its value, which it can define by using a Data Type.

> **Example 1**

Author: I think we should consider it, but generally collapsible stuff is pretty messed up in GitHub markdown I can try and find some examples later but have avoided until now for that reason.

> **Example 1**

Reviewer: This particular example would look something like this:

---

A Property Type can have plain data as its value, which it can define by using a Data Type.

<details><summary>Example 1</summary>
The `Favorite Quote` Property Type could define its value as being an instance of the `Text` Data Type.

- Sample data when used in (simplified view of) an Entity

  ```json
  {
    "favoriteQuote": "He who controls the spice controls the universe"
  }
  ```
</details>

<details><summary>Example 2</summary>
The `Age` Property Type could define its value as being an instance of the `Number` Data Type.

- Sample data when used in (simplified view of) an Entity

  ```json
  {
    "age": 67
  }
  ```
</details>

---

> **Example 1**

Author: You can see problems in the particular example. `code` styled words aren't being rendered properly. I think we might encounter other issues too.

We're also likely going to created rendered pages from the RFCs to host ourselves so for now I think we should keep things very plain markdown to make the generation easier and more consistent. I'd prefer it if we avoided inline HTML because it also decreases readability of the raw MD

> **Example 1**

Reviewer: Okay

- a **required** `$id`, which is a globally unique identifier, where the Property Type's definition can be accessed (in most cases this will be a URL)
- a **required** `name`, which should generally be a non-pluralized description of the property (for example: "Address" not "Addresses")
- **optionally** a `description` to further explain the semantic meaning of the Property Type
- a **required** definition of its possible values

**Defining the possible values of a property**

#### Data Types

A Property Type can have plain data as its value, which it can define by using a Data Type.

**Example 1**

The `Favorite Quote` Property Type could define its value as being an instance of the `Text` Data Type.

- Sample data when used in (simplified view of) an Entity

  ```json
  {
    "favoriteQuote": "He who controls the spice controls the universe"
  }
  ```

**Example 2**

The `Age` Property Type could define its value as being an instance of the `Number` Data Type.

- Sample data when used in (simplified view of) an Entity

  ```json
  {
    "age": 67
  }
  ```

#### Objects

> A Property Type can also have an object made up of other properties as its value. These properties are defined by using Property Types, where when using one it's also possible to define:

Author: I think you're perhaps misunderstanding the phrasing and we should redefine it. It's not saying that talking about the parent Property Type, it's saying it defines the contents of the object through other property types. (as opposed to defining it inline or data types or something)

> A Property Type can also have an object made up of other properties as its value. These properties are defined by using Property Types, where when using one it's also possible to define:

Reviewer: Yeah I think I understand what it's saying, I'm trying to highlight that by saying that the Property Types it's using are distinct (i.e. not inline)

> A Property Type can also have an object made up of other properties as its value. These properties are defined by using Property Types, where when using one it's also possible to define:

Author: Distinct kind of makes it sound like they're unique but technically a property type could have an object as its value which in turn has the same property type.

> A Property Type can also have an object made up of other properties as its value. These properties are defined by using Property Types, where when using one it's also possible to define:

Reviewer: Hadn't considered that referenced property types aren't necessarily distinct. Still don't find the phrasing in this sentence very clear, but can't seem to come up with an alternative so let's leave it for now.

    ```json
    [
      // Person entities
      {
        "entityId": 311,
        "name": "Alice",
      },
      {
        "entityId": 312,
        "name": "Bob",
      }
      {
        "entityId": 313,
        "name": "Charlie",
        "friendsOf": [311, 312] // referring to the Person entity IDs, where the array ordering is unstable
      }
    ]
    ```

- **Specifying there is an Ordered List of Links** - The Entity Type can also express that its out-going links (of the same type) are ordered

  **Example 1**

  The `Playlist` Entity Type could contain some Property Types, and an _ordered_ list of `Contains` links

  - Sample (simplified) data

    ```json
    [
      // Songs
      {
        "entityId": 412,
        "name": "Rocket Man",
        ...
      },
      {
        "entityId": 413,
        "name": "Du Hast",
        ...
      },
      {
        "entityId": 414,
        "name": "Valley of the Shadows",
        ...
      },
      // Playlist
      {
        "entityId": 415,
        "name": "Favorite Songs",
        "contains": [412, 414, 413] // referring to the song entity IDs, ordering is intentional and stable
        ...
      }
    ]
    ```

>     **Example 2**

Author: Yeah I think we need to rework the example formatting in general 🙃 

There are just too many necessary levels of headings and such

>     **Example 2**

Author: Resolving as I don't think this formatting is worth our investment now.

### Structure-based Queries

Queries on the structure of data (those which do not directly reference an exact Entity Type), such as `AggregateEntities`, `AggregateEntityType` and any future methods the Block Protocol supports, behave somewhat similarly to expressing a type. Requiring combinations of properties, constraints, etc. will need to refer to the respective Property Types, Link Types, etc.

### Link Constraints

This RFC does not currently mention ways of constraining the destination of a Link Type used in an Entity Type. **This is different from current way of handling things**. The detailed reasoning for removing this, and a potential solution, is outlined in a [follow-up RFC](https://github.com/blockprotocol/blockprotocol/pull/354).

### Interfacing with Types

The largest change (in terms of the number of interfaces modified) is unsurprisingly the updates that will need to be made to the Entity Type functions.

- The Type objects will be updated to capture the new structures and references between them

> - New `create`, `get`, `update`, and `delete` methods will need to be made for Property Types, Data Types, and Link Types

Author: I think they should probably still have a `get` method but that's about it

- `Number`

  ```json
  {
    "kind": "dataType",
    "$id": "https://blockprotocol.org/types/@blockprotocol/data-type/number",
    "name": "Number",
    "description": "An arithmetical value (in the Real number system)",
    "type": "number"
  }
  ```

- `Boolean`

  ```json
  {
    "kind": "dataType",
    "$id": "https://blockprotocol.org/types/@blockprotocol/data-type/boolean",
    "name": "Boolean",
    "description": "A True or False value",
    "type": "boolean"
  }
  ```

- `Null`

  ```json
  {
    "kind": "dataType",
    "$id": "https://blockprotocol.org/types/@blockprotocol/data-type/null",
    "name": "Null",
    "description": "A placeholder value representing 'nothing'",
    "type": "null"
  }
  ```

- `Object`

  ```json
  {
    "kind": "dataType",
    "$id": "https://blockprotocol.org/types/@blockprotocol/data-type/object",
    "name": "Object",
    "description": "A plain JSON object with no pre-defined structure",
    "type": "object"
  }
  ```

- `Empty List`

  ```json
  {
    "kind": "dataType",
    "$id": "https://blockprotocol.org/types/@blockprotocol/data-type/empty-list",
    "name": "Empty List",
    "description": "An Empty List",
    "type": "array",

>     "maxItems": 0

Author: As the List is empty. There are some other constraints we could do.

Perhaps it'd actually be better to write
```suggestion
    "const": []
```

  }
  ```

### Creating new Data Types

Due to the size of this RFC, creating new Data Types has been deemed out-of-scope and a [follow-up RFC](https://github.com/blockprotocol/blockprotocol/pull/355) is in the works to spec out how this could work, please consult it for plans on introducing a suite of Data Types including ones like `Date`, `Positive Number`, etc.:

> 

Author: I think the reference-level explanation should largely be self-contained. It's mostly a rewrite of the guide-level in much more detail.

As such it's probably worth still having this disclaimer. We could perhaps consider removing it from the guide-level but I think it's better to answer a potential question earlier rather than to distract a reader who might go searching.

          "required": ["oneOf"]
        },
        {
          "properties": {
            "type": { "const": "array" },
            "items": { "$ref": "#/$defs/propertyValues" },
            "minItems": { "type": "number" },
            "maxItems": { "type": "number" }
          },
          "required": ["type", "items"]
        }
      ]
    },
    "propertyTypeObject": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "type": { "const": "object" },
        "properties": {
          "type": "object",
          "propertyNames": {
            "$comment": "Property names must be a valid URI to a Property Type",
            "type": "string",
            "format": "uri"
          },
          "patternProperties": {
            ".*": {
              "oneOf": [
                { "$ref": "#/$defs/propertyTypeReference" },
                {
                  "type": "object",
                  "properties": {
                    "type": { "const": "array" },
                    "items": { "$ref": "#/$defs/propertyTypeReference" },
                    "minItems": { "type": "number" },
                    "maxItems": { "type": "number" },

>                     "additionalProperties": false

Author: I agree, I think we want `"required": ["type", "items"]` and to move `"additionalProperties"` as you say

>                     "additionalProperties": false

Reviewer 2: Will make the change and push in a second

    {
      "$ref": "https://blockprotocol.org/types/@blockprotocol/data-type/number"
    }
  ]
}
```

Sample data when used in (simplified view of) an Entity

```json
{
  "https://blockprotocol.org/types/@alice/property-type/user-id": 42088130893
}
```

or

```json
{
  "https://blockprotocol.org/types/@alice/property-type/user-id": "c09b1839-8084-4a2d-9713-5d074c9c6ce2"
}
```

**Example 2**

The `Contrived Property` Property Type could define its value as being _either_ an instance of the `Number` Data Type, _or_ an object which has a list with a maximum of 4 `Foo` property values

```json
{
  "kind": "propertyType",
  "$id": "https://blockprotocol.org/types/@alice/property-type/contrived-property",
  "name": "Contrived Property",
  "oneOf": [
    {
      "$ref": "https://blockprotocol.org/types/@blockprotocol/data-type/number"
    },
    {
      "type": "object",
      "properties": {
        "https://blockprotocol.org/types/@blockprotocol/data-type/number": {
          "type": "array",
          "items": {
            "$ref": "https://blockprotocol.org/types/@blockprotocol/data-type/number"
          },
          "maxItems": 4
        }
      }

>     }

Author: Good catch

      "author": "David Wilkinson",
      "contents": ..., // omitted for brevity
    }
    ```

- **Specifying there is a list of properties** - The Entity Type can also define whether its properties are lists, where

  - the elements of the list are described by a Property Type
  - there is an optionally specified _maximum_ amount of items in the list
  - there is an optionally specified _minimum_ amount of items in the list

  **Example 1**

  The `Car` Entity Type could contain the Property Types `Make`, `Model`, `Year`, `Brake Horsepower`, and a list of `Extra Trim`

  - Sample (simplified) Data
    Assuming that

    - The `Model`, `Make`, `Spec`, and `Year` Property Types all have values that are instances of the `Text` Data Type
    - The `Extra Trim` Property Type is a list where the list items are instances of the `Text` Data Type
    - The `Brake Horsepower` Property Type has a value that is an instance of the `Number` Data Type

    ```json
    {
      "make": "Mercedes-Benz",
      "model": "300 SL",
      "year": "1957",
      "brake_horsepower": 222,
      "extra_trim": ["Leather Seats", "Cream and Red"]
    }
    ```

#### Defining Relationships between Entities

> Entity Types can also express the types of relationships they have with other things.

Reviewer: Fair, I'll resolve

- a **required** `description` to further explain the semantic meaning of the relationship
- a **possible** `relatedKeywords`, which is a list of terms that are related to the link type

> 💭 As the next section (Entity Types) will outline, there's an additional step to quantifying the types of things on either end of the relationship, as such the descriptions **should** be very general and most of the time not contain specific words about types (such as person).

**Example 1**

There could be an `Owns` link type with a `description` of "Have (something) as one's own; possess"

**Example 2**

There could be a `Submitted By` link type with a `description` of "Suggested, proposed, or presented by"

### Entity Types

An Entity Type is a description of a particular "thing", made up of identifiable pieces of data. An Entity Type is composed of the following:

- a **required** `$id`, which is a globally unique identifier, where the Entity Type's definition can be accessed (in most cases this will be a URL)
- a **required** `name`, which should generally be a non-pluralized description of the thing
- **optionally** a `description` to further explain the semantic meaning of the Entity Type

> - a **required** definition of its possible values.

Author: Feel free to commit, I don't have a strong opinion on this specific phrasing. If you think it improves readability please add it, the following is just more of a thought stream in general on this.

I think generally speaking it might be helpful for us to use "values" more generally than you seem to associate it with. E.g. the "value" of a Property _Type_, etc. (similar to how you used to use "point to" when coming more from a property graph view of things)

And I do think the values of an Entity are defined by the Property Types in the Entity Type, it's just that those Property Types then in turn define their values by Data Types (and other stuff). It all gets collapsed.

> - a **required** definition of its possible values.

Reviewer: I personally don't find the idea of a "value" of an entity type very clear, so think being more specific would be helpful here so will commit it. Better understand where you're coming from now though, thanks.

{
  "$id": "https://blockprotocol.org/type-system/0.2/schema/meta/entity-type",
  "description": "Specifies the structure of an Entity Type",
  "properties": {
    "kind": {
      "const": "entityType"
    },
    "$id": {
      "type": "string",
      "format": "uri"
    },
    "name": { "type": "string" },
    "description": { "type": "string" },
    "properties": { "$ref": "#/$defs/propertyTypeObject" },
    "required": {
      "type": "array",
      "items": { "type": "string" }
    },
    "requiredLinks": {
      "$comment": "A list of link-types which are required. This is a separate field to 'required' to avoid breaking standard JSON schema validation",
      "type": "array",
      "items": { "type": "string" }
    },
    "links": {
      "type": "object",
      "propertyNames": {
        "$comment": "Property names must be a valid URI to a link-type",
        "type": "string",
        "format": "uri"
      },
      "patternProperties": {
        ".*": {
          "type": "object",
          "oneOf": [
            {
              "properties": {
                "ordered": { "type": "boolean", "default": false },
                "type": { "const": "array" }
              },

>               "required": ["ordered", "type"]

Reviewer: Sorry if I'm misunderstanding something, but what purpose does a default value serve when the schema developer is forced to specify the field's value explicitly? When would the default value be used?

>               "required": ["ordered", "type"]

Author: > The default keyword specifies a default value. This value is not used to fill in missing values during the validation process. Non-validation tools such as documentation generators or form generators may use this value to give hints to users about how to use a value. However, default is typically used to express that if a value is missing, then the value is semantically the same as if the value was present with the default value. The value of default should validate against the schema in which it resides, but that isn’t required.

- https://json-schema.org/understanding-json-schema/reference/generic.html

We would use it to inform the default choice in the schema editor, for example.

>               "required": ["ordered", "type"]

Reviewer 2: Oh I thought it was used during the validation process. Nevermind then!

>               "required": ["ordered", "type"]

Author: Worth highlighting that annotations are not functional during validation by default (to my understanding)

>               "required": ["ordered", "type"]

Reviewer: Ah that makes sense, thanks for sharing

A key change for allowing the proposed type system to work is moving away from arbitrary property keys and making use of canonical property URIs. As seen in the examples in earlier sections, properties that use simple keys will now have to point at _Property Type URIs_.

**An example of an entity instance in the current system:**

```json
{
  "entityId": 111,
  "name": "Arthur Philip Dent",
  "age": 30,
  "planetOfOrigin": "Earth",
  "occupation": "Intergalactic Traveler"
}
```

**An example of an entity instance in the proposed system:**

```json
{
  "https://blockprotocol.org/types/@blockprotocol/property-type/entity-id": 111,
  "https://blockprotocol.org/types/@alice/property-type/name": "Arthur Philip Dent",
  "https://blockprotocol.org/types/@alice/property-type/age": 30,
  "https://blockprotocol.org/types/@alice/property-type/planet-of-origin": "Earth",
  "https://blockprotocol.org/types/@alice/property-type/occupation": "Intergalactic Traveler"
}
```

This change canonicalizes property keys, such that they uniquely identify a Property Type.

**Within blocks** this impacts

- how entity fields are accessed
- how Block Protocol messages are constructed

> - payload size of Block Protocol messages

Reviewer 2: yes, it refers to that

**For block authors** this impacts:

- how Block Schemas are defined
- reusability of properties (the need for some kind of type discovery)

> 💡 Notice that `entityId`s are replaced by a Property Type URI that resides in the `/@blockprotocol` namespace. This is an implementation detail, and not something that is strictly dictated by the proposal.

### Interfacing with Linked Entities

#### Receiving Links and Linked Entities

Link groups and Linked entities in the Graph Service are currently supplied outside the entity as separate objects. This behavior will stay the same, but the objects received will be of a different shape.

> **An example of a `linkGroup` instance in the current system:**

Reviewer 2: It would be the item in the `linkGroups` array, yes, but the type there is `$ref`d to be [/graph/link-group](https://blockprotocol.org/types/services/graph/link-group) in [the spec](https://github.com/blockprotocol/blockprotocol/blob/main/packages/%40blockprotocol/graph/src/graph-service.json#L773) which is what's being referred to here. 

I think it would be really nice to be able to link directly to these directly, I was just unaware that the types were live yet. I might go through the examples in this section and add direct links to them

> **An example of a `linkGroup` instance in the current system:**

Reviewer 2: Unrelated, i think the `current` example is wrong here, whoopsie!

```json
{
  "sourceEntityId": "user1",
  "path": "company",
  "ordered": false,
  "links": [
    {
      "sourceEntityId": "user1",
      "destinationEntityId": "company1",
      "path": "company"
    }
  ]
}
```

**An example of a `linkGroup` instance in the proposed system:**

```json
{
  "sourceEntityId": 111,
  "link": "https://blockprotocol.org/types/@alice/link-type/company",
  "ordered": false,
  "links": [
    {
      "sourceEntityId": 111,
      "destinationEntityId": 222,
      "link": "https://blockprotocol.org/types/@alice/link-type/company"
    }
  ]
}
```

> > 💡 In the proposed system example instance, the `path` key has been replaces with `link`. This is an implementation detail which is not dictated by the proposal. Alternatives could be `linkUri`, `linkType`, etc.

Author: We can do, but it's also fine (and less hassle) to decide on the naming of some keys when actually implementing things 🤷

                  "required": ["type", "items"],
                  "additionalProperties": false
                }
              ]
            }
          }
        },
        "required": {
          "type": "array",
          "items": "string"
        }
      }
    },
    "propertyTypeReference": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "$ref": {
          "$comment": "Property Object values must be defined through references to the same valid URI to a Property Type",
          "type": "string",
          "format": "uri"
        }
      },
      "required": ["$ref"]
    },
    "dataTypeReference": {
      "type": "object",
      "additionalProperties": false,
      "properties": {
        "$ref": {
          "type": "string",
          "format": "uri"
        }
      },
      "required": ["$ref"]
    }
  }
}
```

### Defining the possible values of a property

A Property Type describes its possible values through a `oneOf` definition, which can include:

#### Data Types

**Example 1**

The `Favorite Quote` Property Type could define its value as being an instance of the `Text` Data Type.

```json
{
  "kind": "propertyType",

>   "$id": "https://blockprotocol.org/types/@alice/property-type/favorite-quote",

Author: Sticking the comments here @nonparibus so we can reply in a chain because commenting directly on the PR is a mess (thanks GitHub 🙃 )

> 
> General nit 1: it's tautological to have both `types` and `[type]-type` in the URL slug here. Let's simplify `data-type` to `data`, along with the other types.
> 
> General nit 2: the namespace should come before `types` in the slug to match the public site.

> In this example, the new slug would be `https://blockprotocol.org/@blockprotocol/types/data/number`
> 
> In addition to `data`, orgs would also have:
> - `https://blockprotocol.org/@blockprotocol/types/entity`
> - `https://blockprotocol.org/@blockprotocol/types/link`
> - `https://blockprotocol.org/@blockprotocol/types/property`

Author: The URLs have always meant to be placeholders (although included to be indicative that URIs are verbose).

Thanks for highlighting. When it comes to the implementation we should decide on appropriate URI patterns

Reviewer 3: Apologies @Alfred-Mountfield, @benwerner01 has previously flagged this weird GH behavior to me. My bad.

Reviewer 3: You think that would be better than...
- `https://blockprotocol.org/@namespace/type/data/...`
- `https://blockprotocol.org/@namespace/type/property/...`
- `https://blockprotocol.org/@namespace/type/entity/...`
- `https://blockprotocol.org/@namespace/type/link/...`

Author: I'm sort of inclined to go for something else, where I don't have a concrete suggestion yet. My thought is that we probably don't want service-specific things just living in plain paths on blockprotocol.org. If this RFC gets merged with the Type System _as part of the `graph` service_, then I think we should probably look at having `graph` or `service/graph` or something in the path as well.

I do much prefer `data-type` over `types/data` too.

Reviewer 2: The `/type/property/` pattern feels a little like browsing a tree (sort of a file system), which could imply that there will be more kinds after the `/types/` section.

The `/entity-type/` is more clear about the limited kinds we're dealing with.

I don't have a strong preference about what the naming should look like, just sharing my initial thoughts on the two variants.

Reviewer: @nonparibus the benefit I see is whether we're talking about data (entities, properties, etc) or types (entity-types, property-types, etc) the URI has a consistent structure:
- `https://blockprotocol.org/[namespace]/[what is it]/[identifier]`
- `https://blockprotocol.org/@alice/entity-type/Person`
- `https://blockprotocol.org/@alice/entity/Elon-Musk`

But I don't have strong feelings about this so happy to go either way

Author: Yes but to be fair `[what is it]` doesn't **have** to be a single part of the path, it could be a few things like `types/entity` or `services/graph/entity-type` or something else

Reviewer 3: I think having `service/graph` would be overkill, because we can assume that any merged thing represented in the spec is part of the core or a `service`.

I am arguing for `https://blockprotocol.org/[namespace]/[what is it]/[identifier]` wherein `[what is it]` is simply `type/:type:`, because we have types _of_ types (entity/property/link/data). I do not care too much if we think `entity-type` is better than `type/entity`, although because the latter better represents a _type of type_ it seems to provide more descriptive information per-bit (at least assuming my conception of types is correct).

Author: I think there is some value in considering namespacing according to the service so that we don't have to worry about overlapping things, (perhaps some event service comes up with a need for `type` or whatever). I don't feel particularly strongly about it.

I don't think _type of types_ is incorrect but perhaps its easiest for us to call them "classes of types".

My main thing about `entity-type` v `types/entity` mostly stems from a somewhat silly opinion that `types/data` looks ugly and I wouldn't really want us to ever call a Data Type just "Data" because it could easily get very confusing.

Author: I think we're going to leave them as is for this RFC, and address properly in the near future when we specify how URIs are going to be formatted. This will possibly be done as part of #408

A **Data Type** describes a space of possible valid _values_. For instance, the _string_ value `"foo"` might be an instance of a `Text` Data Type, or a _float_ value `0.6` might be an instance of a `Number` Data Type ("instance of" means that the value satisfies the constraints expressed on the Data Type, or alternatively you can view it as saying the value is within the value space of the Data Type).

> 💬 The underlying value's representation will depend on the implementation environment, different languages and technologies may store these differently.

A Data Type is composed of the following:

- a **required** `$id`, which is a globally unique identifier, where the Data Type's definition can be accessed (in most cases this will be a URL)
- a **required** `name`, which should generally be a non-pluralized short description of the Data Type (for example: "Number" not "Numbers")
- **optionally** a `description` to further explain the semantic meaning of the Data Type
- a **required** definition of its possible value

#### Primitive Top-Level Data Types

We pre-define a set of primitive top-level Data Types.
These can be considered the least restrictive value spaces:

- `Text`

> - `Number`

Author: Should we include `Integer` here? Or save that for user-created ones?

Reviewer: My thinking is `Integer` is a more constrained version of `Number`, therefore should not be considered as one of our "primitive" data types. So yes, we should probably introduce it as part of the "user-created" data types.

Author: I'm inclined to agree, my only worry is that if user-generated types does take a while to be merged, then users can't actually specify an integer type until then.

- Feature Name: graph-type-system
- Start Date: 2022-06-14
- RFC PR: [blockprotocol/blockprotocol#352](https://github.com/blockprotocol/blockprotocol/pull/352)
- Block Protocol Discussion: [blockprotocol/blockprotocol#0000](https://github.com/blockprotocol/blockprotocol/discussions/0000)

<!-- markdownlint-disable  MD036 -->

# Summary

[summary]: #summary

This RFC proposes changing the way data structural requirements are expressed within the BP.

Rather than allowing any free-form JSON schema, this RFC defines the following classes of constrained JSON schemas:

- Data Types
- Property Types
- Link Types
- Entity Types

These elements are used in combination to improve reusability of, and better define, expressions of data structural requirements.

The structure of these types are defined by (and can be validated against) their meta-schemas, which can be found in the [reference-level explanation](#reference-level-explanation).

> By using the proposed types, blocks, embedding applications, and end-users can all have greater confidence in each other's roles and functionalities, and hopefully will allow them all to benefit from better computational inference and capabilities. Reusability of descriptions of data should decrease developmental overhead over time, with all members of the ecosystem benefiting from emergent standardization.

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/83d7bee49c445a452b4b4c9181fe075c17976da5. As this is already a large review, I've tried not to go too much into depth

```json
{
  "display_picture": "https://dummyimage.com/300.jpg",
  ... // some other fields
}
```

This is an example of a lack of **consensus**, where two parties don't agree on how to describe the data in question. At this point, there isn't a clear route to deciding compatibility. Block schemas could perhaps be modified so that the block could express it needs something that matches a [regex to ensure the URL has an image extension](https://stackoverflow.com/questions/42853961/regex-that-matches-image-links-but-exclude-normal-url), and the embedding application could ignore the key and just match on values, but this:

- is potentially flakey ([here's an example](https://dummyimage.com/300) of a URL that points to an image but _does not_ have an extension)
- ends up being very domain-specific
- introduces a large maintenance burden
- is generally not reusable, possibly leading to similar bespoke logic being defined in multiple places, with possible subtle but incompatible variations

Thus the main reason for following RFC:

> > 💬 Removing some flexibility to establish a stronger contract between the block developer, embedding application, and end-user, allows for more powerful run-time inference and general confidence of all parties. Furthermore, creating a well-defined hierarchy of types, that are reusable and discoverable, encourages emergent standardization and, in turn, leads to higher-quality interoperable parts.

Author: I'm not sure I agree as strongly with this one. People do not have the option of avoiding the Type System under this RFC, as such there is a contract being established that is more constrained (i.e. stronger, perhaps not _better_). Perhaps "stronger contract" has connotations that are too firmly implying direct benefit?

Also regarding the inclusion of "which" I don't _think_ that's grammatically correct, but it's generally because the sentence is very wordy and runs on.
> Removing some flexibility ..., allows for more powerful ...

vs 

> Removing some flexibility ..., which allows for more ....

I think general confidence kind of encompasses more of the subtle benefits of the approach, but perhaps this is too hand-wavy and overplays the benefits. In that case perhaps the general way you're leaning here with the review, of picking more specific key benefits, makes more sense (i.e. in this scenario focusing on improvements in data compatibility). I'm open to suggestion, if you feel remotely strongly I'm happy to commit that :)

> > 💬 Removing some flexibility to establish a stronger contract between the block developer, embedding application, and end-user, allows for more powerful run-time inference and general confidence of all parties. Furthermore, creating a well-defined hierarchy of types, that are reusable and discoverable, encourages emergent standardization and, in turn, leads to higher-quality interoperable parts.

Reviewer 4: I don't feel particularly strongly. I think this would be addressed by making the 'a lot of benefits are unlocked by people sharing types, not just introducing this new system' somewhere.

For what it's worth I think there are two sorts of contracts between block and app introduced by the type system:
1. We agree to all define our schemas in this composable/re-usable way (they must do this)
2. Where we refer to the same schema (they might do this), it means the same thing

> > 💬 Removing some flexibility to establish a stronger contract between the block developer, embedding application, and end-user, allows for more powerful run-time inference and general confidence of all parties. Furthermore, creating a well-defined hierarchy of types, that are reusable and discoverable, encourages emergent standardization and, in turn, leads to higher-quality interoperable parts.

Author: I agree with those points, although I'm not sure how much time we want to spend refining this technical document to capture all of those things (not saying you're suggesting we work them in).

As an action item though we'll add a clear section near the top about how most of the benefits are basically gated by people participating in an ecosystem and sharing things.

> > 💬 Removing some flexibility to establish a stronger contract between the block developer, embedding application, and end-user, allows for more powerful run-time inference and general confidence of all parties. Furthermore, creating a well-defined hierarchy of types, that are reusable and discoverable, encourages emergent standardization and, in turn, leads to higher-quality interoperable parts.

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/555dd097384f416a42afacfe001870a98ed096e0, made it short-ish as to not spend too much time refining the contents

> In the following document we outline an approach to being more specific about the types that describe data. Through the following definitions of types, the community will be able to **gradually converge** upon a **consensus** of how to describe information.

Reviewer 4: Best I can do
```suggestion
In the following document we outline an approach to being more specific about the types that describe data. Through the following definitions of types, the community will be more likely to **gradually converge** upon a **consensus** of how to describe information.
```

> In the following document we outline an approach to being more specific about the types that describe data. Through the following definitions of types, the community will be able to **gradually converge** upon a **consensus** of how to describe information.

Author: I don't love it either but it's better than the garbage I was coming up with, and it's better than it was in my opinion. Thanks!

## Motivating implications

**embedding applications:**

- are required to do less guesswork about what the block is asking for
- require less user intervention (or hard-coded logic/guesses) to map X to Y

> **blocks:**

Author: Firstly, I 100% agree with the issue you've identified, I think we should be a lot more up-front and clear about how the potential is mostly only realised if we can guide people into re-using types, etc.

I don't fully agree that there isn't value without that happening though, I think a lot of the bullet points are still true even if there isn't common overlap in the types between EAs and Blocks, because this RFC gives more ways for mismatched descriptions to be addressed. It is easier for an EA to map disjoint descriptions of "image" to one when they have things (types) to use, and more information (embedded in those types) available to make the decision.

> **blocks:**

Reviewer 4: What more information is carried by a separate property definition compared to an in-line property definition? 

How does an EA map between two different property definitions without user intervention?

> **blocks:**

Author: > What more information is carried by a separate property definition compared to an in-line property definition?

Property Types require more information than we were previously asking for in inline things. For one, they will be associated with Data Types which later on _may_ encapsulate more semantic meaning, but let's disregard that as it's not part of this RFC. The main thing is that in this case Property Types are matched up with descriptions, and we can make it possible to see the type of contexts they're used in (by showing blocks that use them in the schemas for example). This type of contextualisation is not possible/enforced when your property is just a key. 

Using URIs as identifiers also lets us (even programmatically if such things are stored) disambiguate between properties that have the same name but different semantic meaning. 

Now a lot of it isn't really able to be programmatically inferred on first discovery, so I don't think some super general EA will look at the `description` of `image_url` and know that it's on about a profile picture or something. However for EAs that are including types in the source-code, or ones that are storing user decisions, this proposal provides new routes. i.e. if I've seen `image_url` before and the user has been able to encapsulate the semantic meaning in a way that I can store, when I see another block using the same property type I can reuse that stored meaning, as opposed to just seeing another string that's equivalent and having no idea if it's on about the same thing.

> How does an EA map between two different property definitions without user intervention?

I make no claims about the implementation details here, or that there won't be user intervention. I'm only suggesting that there are implementation possibilities that are now available that weren't before. If a _user stores_ some mapping of property definitions, then that can still be reused without user intervention if the property appears in some previously unseen block, etc.

> **blocks:**

Author: You have swayed me, thank you :D I am adding the suggestion. I think there are benefits without type sharing but I do not think the bullet points are the ones. As you pointed out all of my examples (regarding programmatic things) actually require reuse at some point!

- have greater confidence that the data passed to them is correct and useable
- are able to be more specific about their requirements and more easily express the function of the data

>   - This also facilitates more powerful composition as there's greater confidence about other parties' functionalities.

Reviewer 2: It refers to `the function of the data`, perhaps using the word semantics would be more sufficient

<!-- markdownlint-disable  MD029 -->

3.  define a constrained type system that is able to describe any JSON object structure
4.  refine the current way of defining links between entities, encapsulating the approach within the type system

<!-- markdownlint-enable  MD029 -->

## Types

Underpinning this proposal is the definition of four classes of "types". One of these classes is an extension of the current definition of an _Entity Type_ while the other three are new classifications that are designed to improve the usability and expressiveness of _Entity Types_.

### Data Types

A **Data Type** describes a space of possible valid _values_. For instance, the _string_ value `"foo"` might be an instance of a `Text` Data Type, or a _float_ value `0.6` might be an instance of a `Number` Data Type ("instance of" means that the value satisfies the constraints expressed on the Data Type, or alternatively you can view it as saying the value is within the value space of the Data Type).

> 💬 The underlying value's representation will depend on the implementation environment, different languages and technologies may store these differently.

A Data Type is composed of the following:

> - a **required** `$id`, which is a globally unique identifier, where the Data Type's definition can be accessed (in most cases this will be a URL)

Author: Oops so we don't, will fix that! Cheers for spotting

> - a **required** `$id`, which is a globally unique identifier, where the Data Type's definition can be accessed (in most cases this will be a URL)

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/227b7b219ad2f67324cf03e63f8685c5ff564a73, thanks!

> - a **required** `name`, which should generally be a non-pluralized short description of the Data Type (for example: "Number" not "Numbers")

Author: > that we use the 'name' property to refer to these (and other types of types)

It's probably worth us figuring out what we want to use `name` for, it's included because it seems helpful for acting as a display name, but perhaps that isn't the only purpose. We should find a way to be clearer that it is not guaranteed to be unique across types, I definitely agree with that.

> if there must or should be any relation between $id and name

We can look at establishing convention for this, but in the mean-time I'm not sure I see genuine value in enforcing that there is some relationship between them. I don't think the URI _**has**_ to include the `name` for it to still be useful and valid, but perhaps we should suggest it usually should. (this gets especially tricky if we enter a world where people can rename things)

> - a **required** `name`, which should generally be a non-pluralized short description of the Data Type (for example: "Number" not "Numbers")

Reviewer 4: I don't think we need to enforce a relationship between the two. When we come to specify we should explicitly say that because I expect our examples will typically imply some relation between them (because the `$id` invariably ends in a kebab case version of the name)

- **optionally** a `description` to further explain the semantic meaning of the Data Type

> - a **required** definition of its possible value

Author: This is purposefully (and perhaps incorrectly) left unclear here due to the needed follow-up RFC for #355.

`EmptyList` is an example where it's not `type`, and we expect to see many more constraints appearing, for example `format`, `pattern`, etc.

> - a **required** definition of its possible value

Reviewer 4: It must _include_ `type` in all those instances, but I appreciate the nuance. Resolving.

### Property Types

A **Property Type** is a description of a named piece of data, including its possible values. A Property Type is composed of the following:

- a **required** `$id`, which is a globally unique identifier, where the Property Type's definition can be accessed (in most cases this will be a URL)
- a **required** `name`, which should generally be a non-pluralized description of the property (for example: "Address" not "Addresses")
- **optionally** a `description` to further explain the semantic meaning of the Property Type
- a **required** definition of its possible values

**Defining the possible values of a property**

#### Data Types

A Property Type can have plain data as its value, which it can define by using a Data Type.

**Example 1**

The `Favorite Quote` Property Type could define its value as being an instance of the `Text` Data Type.

- Sample data when used in (simplified view of) an Entity

  ```json
  {

>     "favoriteQuote": "He who controls the spice controls the universe"

Author: I was worried that the simplification would be confusing, thank you for confirming.

Could we fix this by adding a big ol' disclaimer saying that the examples in this section are not representative and use the name and blah blah blah?

I think there's still value in understanding how these types describe data, and how it isn't _too_ different from the current way of doing things beyond people needing to rewrite keys. I'm not sure how I feel about not giving those high-level explanatory examples in this section and forcing people to be faced with the Reference-Level which really is a lot less approachable.

If a disclaimer wouldn't really work perhaps there's something else we can do, if you think the data examples don't actually give much value I'm content with dropping them.

>     "favoriteQuote": "He who controls the spice controls the universe"

Reviewer 4: I think using non-concrete URIs would help.

> Could we fix this by adding a big ol' disclaimer saying that the examples in this section are not representative and use the name and blah blah blah?

I don't think this will help because the examples do not use the name. You could say 'the examples use a camel case version of the name', but then it becomes even more explicitly 'Here are some examples of things which we aren't proposing'

>     "favoriteQuote": "He who controls the spice controls the universe"

Reviewer 2: Added non-concrete URIs in https://github.com/blockprotocol/blockprotocol/pull/352/commits/30d59ba15a0e4c415fa25f065be90aa306380004

    - The `Extra Trim` Property Type is a list where the list items are instances of the `Text` Data Type
    - The `Brake Horsepower` Property Type has a value that is an instance of the `Number` Data Type

    ```json
    {
      "make": "Mercedes-Benz",
      "model": "300 SL",
      "year": "1957",
      "brake_horsepower": 222,
      "extra_trim": ["Leather Seats", "Cream and Red"]
    }
    ```

#### Defining Relationships between Entities

Entity Types can also express the types of relationships they have with other things.

- **Specifying an Entity Type has outgoing Links** - The links going _from_ an Entity are described a set of Link Types

  **Example 1**

  The `Book` Entity Type could contain the Property Types outlined above, and a `Written By` link

  - Sample (simplified) data with the same assumptions as outlined above

    ```json
    [
      // Person entity
      {
        "entityId": 111,
        "name": "Herbert George Wells",
        ...
      },
      // Book Entity
      {
        "entityId": 112,
        "name": "The Time Machine",
        "publishedOn": "1895-05",
        "blurb": ...,

>         "writtenBy": 111 // referring to the Person entity ID

Author: Was also worried the simplification would be confusing here. I think it depends on your view of the world but we should probably try and align with the intuition that may be established with familiarity of the BP graph methods (in that as you say links are returned separately).

We can rework these.

>         "writtenBy": 111 // referring to the Person entity ID

Reviewer 2: I've went through all examples and explicitly split out `entityId`, `properties` and `links` and added a disclaimer in the BP implications section.

In https://github.com/blockprotocol/blockprotocol/pull/352/commits/f19259a56162d6244ed29416cfeb165217312a06

#### Creating Links

The Link Functions will be updated to handle Link Types, removing the `path` fields and instead use the URI of the Link Type.

### Block Schemas

Block schemas will need to be updated to be built using the types outlined above. The specifics of this are explored in-depth in the Reference-Level explanation, however the basic premise is that instead of writing new definitions inline in every block schema, the properties and links should be defined as references to their respective types. If a type does not exist for a given property then it will need to be created, and so on.

### Structure-based Queries

Queries on the structure of data (those which do not directly reference an exact Entity Type), such as `AggregateEntities`, `AggregateEntityType` and any future methods the Block Protocol supports, behave somewhat similarly to expressing a type. Requiring combinations of properties, constraints, etc. will need to refer to the respective Property Types, Link Types, etc.

### Link Constraints

> This RFC does not currently mention ways of constraining the destination of a Link Type used in an Entity Type. **This is different from current way of handling things**. The detailed reasoning for removing this, and a potential solution, is outlined in a [follow-up RFC](https://github.com/blockprotocol/blockprotocol/pull/354).

Author: Going to leave this thread open as it's an important point. Perhaps we go with your prior suggestion and continue to allow pointing to an exact type, and address the problem with the other RFC rather than introducing some sort of implementation dependency.

```json
{
  "kind": "propertyType",
  "$id": "https://blockprotocol.org/types/@alice/property-type/numbers",
  "name": "Numbers",
  "oneOf": [
    {
      "type": "array",
      "items": {
        "oneOf": [
          {
            "$ref": "https://blockprotocol.org/types/@blockprotocol/data-type/number"
          }
        ]
      }
    }
  ]
}
```

The `Numbers` Property Type models data of the following shape

```json
[1, 2, ...]
```

#### Describing multiple possibilities

A Property Type can also express that its value is _either_ something _or_ something else, where "something" and "something else" are defined as outlined above using Data Types or an object definition.

**Example 1**

The `User ID` Property Type could define its value as being _either_ an instance of the `Text` Data Type _or_ an instance of the `Number` Data Type.

```json
{
  "kind": "propertyType",
  "$id": "https://blockprotocol.org/types/@alice/property-type/user-id",
  "name": "User ID",
  "oneOf": [
    { "$ref": "https://blockprotocol.org/types/@blockprotocol/data-type/text" },
    {
      "$ref": "https://blockprotocol.org/types/@blockprotocol/data-type/number"
    }
  ]
}
```

> Sample data when used in (simplified view of) an Entity

Author: Very good spot, this wasn't a conscious decision and it's not something we meant to implicitly propose, this was mostly due to the timings around 0.2 and this RFC being re-written, we should update this to fix it.

> Sample data when used in (simplified view of) an Entity

Reviewer 2: I've updated all examples to be more in line with BP 0.2, and also made it clear in the implications section that the examples that contain both `properties` and `links` will change shape when used in BP.

Addressed in https://github.com/blockprotocol/blockprotocol/commit/f19259a56162d6244ed29416cfeb165217312a06

  "required": ["kind", "$id", "name", "description"]
}
```

> 💭 As the next section (Entity Types) will outline, there's an additional step to quantifying the types of things on either end of the relationship, as such the descriptions **should** be very general and most of the time not contain specific words about types (such as person).

**Example 1** - `Owns`

```json
{
  "kind": "linkType",
  "$id": "https://blockprotocol.org/types/@alice/link-type/owns",
  "name": "Owns",
  "description": "Have (something) as one's own; possess"
}
```

**Example 2** - `Submitted By`

```json
{
  "kind": "linkType",
  "$id": "https://blockprotocol.org/types/@alice/link-type/submitted-by",
  "name": "Submitted By",
  "description": "Suggested, proposed, or presented by"
}
```

## Entity Types

An **Entity Type** is a JSON schema that satisfies the following JSON meta-schema:

```json
{
  "$id": "https://blockprotocol.org/type-system/0.2/schema/meta/entity-type",
  "description": "Specifies the structure of an Entity Type",
  "type": "object",
  "properties": {
    "kind": {
      "const": "entityType"
    },
    "$id": {
      "type": "string",
      "format": "uri"
    },

>     "name": { "type": "string" },

Author: We started off with `title` but I'm not 100% convinced yet that it is the same usage, perhaps it is.

The `title` property is the name of the **schema**, it's specifically about what the schema does, in the same way that we have to be careful about our usages of `description` because describing the type _might not be_ the same as describing the schema of the type.

If we can convince ourselves that they're the same then I agree we should use `title`.

>     "name": { "type": "string" },

Reviewer 4: Here is the [latest JSON Schema definition](https://json-schema.org/draft/2020-12/json-schema-validation.html#name-title-and-description) for `title` and `description`:
> Both of these keywords can be used to decorate a user interface with information about the data produced by this user interface. A title will preferably be short, whereas a description will provide explanation about the purpose of the instance described by this schema.

Seems wide enough to accommodate what we want, but I haven't been in the discussion where you switched away from `title` so may be some nuance to `name` I'm missing.

>     "name": { "type": "string" },

Author: I haven't come up with any good reason for us to not use `title` so we'll switch. Thanks for the suggestion!

>     "name": { "type": "string" },

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/644bfc128e2b0c55e907edff429e6dfa3843dab3, should be updated everywhere where name is mentioned in the context of the schema `name`

    "description": { "type": "string" },
    "properties": { "$ref": "#/$defs/propertyTypeObject" },
    "required": {
      "type": "array",
      "items": { 
        "type": "string",
        "format": "uri"
      }
    },
    "requiredLinks": {
      "$comment": "A list of link-types which are required. This is a separate field to 'required' to avoid breaking standard JSON schema validation",
      "type": "array",
      "items": { "type": "string" }
    },
    "links": {
      "type": "object",
      "propertyNames": {
        "$comment": "Property names must be a valid URI to a link-type",
        "type": "string",
        "format": "uri"
      },
      "patternProperties": {
        ".*": {
          "type": "object",
          "oneOf": [
            {
              "properties": {
                "ordered": { "type": "boolean", "default": false },
                "type": { "const": "array" }
              },
              "required": ["ordered", "type"]
            },
            {}
          ],
          "additionalProperties": false
        }
      }
    }
  },

>   "additionalProperties": false,

Author: It's only half intentional. I think the intention was somewhat that we'd rediscover some of these in implementation and update as necessary. Inheritance and such specifically we know needs a different RFC / explanation somewhere, so rather than blocking this on that it seemed sensible to update these meta-schemas in that explanation.

We should probably add `default` and `examples` in this RFC though.

> We may also wish to include $schema with a const on each type, to point to our meta-schema for that type.

This is something we needed to look into a little further as JSON schema documentation on `$schema` is unclear, and they often have them as self-describing in tutorials (i.e. the `$schema` tag points to `$id`, _not just for the official drafts of JSON schema_), the intention and hope was to do as you suggest, and I think we can.

>   "additionalProperties": false,

Reviewer 4: Fair enough – we don't talk about inheritance in the current spec, so I think it's fine to leave out for now and handle properly when we figure it out.

>   "additionalProperties": false,

Reviewer 2: I think it makes a lot of sense to add both `default` and `examples`. I do have some questions, however:

For both, how do we handle `link`s? 
In this proposal, they are co-located next to the `properties`, but in BP we would split them up. Would we then only allow `examples` and `default` to describe the `properties` part of an Entity Type? Specifying a default for an entity type that requires a link seems impossible

>   "additionalProperties": false,

Reviewer 4: Yes, they would only provide an example value for `properties`. Dummy data involving _other_ records (whether types, other entities, or links) can be provided via `example-graph.json` - [spec](https://blockprotocol.org/docs/spec/graph-service-specification#block-package) - [example](https://github.com/hashintel/hash/blob/main/packages/blocks/table/example-graph.json)

>   "additionalProperties": false,

Reviewer 2: I've added `default` and `examples` to the spec here https://github.com/blockprotocol/blockprotocol/pull/352/commits/75d3277cc5a05689bad0f80bde0bfa9facb40f5c
For now I haven't added any example usages in the RFC or written about it. I think that's fine for now.

            "format": "uri"
          }
        }
      },
      "required": ["type", "properties"],
      "additionalProperties": false
    },
    "propertyTypeReference": {
      "type": "object",
      "properties": {
        "$ref": {
          "$comment": "Property Object values must be defined through references to the same valid URI to a Property Type",
          "type": "string",
          "format": "uri"
        }
      },
      "required": ["$ref"],
      "additionalProperties": false
    }
  }
}
```

### Defining an Entity Type's Properties

#### Specifying an Entity Type has Property Types

The data _within_ an Entity is described simply through a set of Property Types

**Example 1**

The `Book` Entity Type could contain the Property Types `Name`, a `Published On` and a `Blurb`, where `Name` is **required**.

Assuming that:

- the `Name`, `Published On`, and `Blurb` Property Types all have values that are instances of the `Text` Data Type

```json
{
  "kind": "entityType",
  "$id": "https://blockprotocol.org/types/@alice/entity-type/book",
  "type": "object",
  "name": "Book",
  "properties": {
    "https://blockprotocol.org/types/@alice/property-type/name": {

>       "$ref": "https://blockprotocol.org/types/@alice/property-type/name"

Author: We've been trying to figure that out as well, I'm not sure if there's a way we can constrain that in JSON schema itself, it is unresolved question 3.

If we can't do that then perhaps we can establish our own vocabulary for `$ref` which tends to be how JSON schema adds semantic meaning to keywords, the last fallback is we should just add a paragraph somewhere expressing its necessity and saying that validator implementations should check it.

>       "$ref": "https://blockprotocol.org/types/@alice/property-type/name"

Reviewer 4: I think we need it to be in any case, don't we? Otherwise you cannot determine from data instances alone which property they refer to (since the keys can differ).

i.e. we should have a paragraph expressing its necessity now, which notes that we'd ideally have this validated automatically. 

It's a pretty fundamental requirement of realising the benefits of the type system, regardless of how we end up enforcing it.

>       "$ref": "https://blockprotocol.org/types/@alice/property-type/name"

Author: Sorry I didn't mean to imply that we don't need to specify this, we do! I was just saying it'd be good if we could encapsulate it more rigorously than a paragraph. However in lieu of a better solution we shall do so now :)

>       "$ref": "https://blockprotocol.org/types/@alice/property-type/name"

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/9108187fc5d6c497330a677cf22c37d4f381cb74, I've added it to the JSON Schema additions sections rather than adding even more information around the meta schemas.

      "$ref": "https://blockprotocol.org/types/@alice/property-type/published-on"
    },
    "https://blockprotocol.org/types/@alice/property-type/blurb": {
      "$ref": "https://blockprotocol.org/types/@alice/property-type/blurb"
    }
  },
  "links": {
    "https://blockprotocol.org/types/@alice/property-type/written-by": {}
  },
  "requiredLinks": [
    "https://blockprotocol.org/types/@alice/property-type/written-by"
  ]
}
```

This would accept Entity instances with the following shape

```json
[
  // Person entity
  {
    "https://blockprotocol.org/types/@blockprotocol/property-type/entity-id": 111,
    "https://blockprotocol.org/types/@alice/property-type/name": "Herbert George Wells",
    ...
  },
  // Book Entity
  {
    "https://blockprotocol.org/types/@blockprotocol/property-type/entity-id": 112,
    "https://blockprotocol.org/types/@alice/property-type/name": "The Time Machine",
    "https://blockprotocol.org/types/@alice/property-type/published-on": "1895-05",
    "https://blockprotocol.org/types/@alice/property-type/blurb": ...,
    "https://blockprotocol.org/types/@alice/property-type/written-by": 111 // referring to the Person entity ID
  }
]
```

> #### Specifying there is a List of Links

Author: Again very good spot! I thought we had done so, we will fix that.

    "enum": ["entityType", "propertyType", "dataType", "linkType"]
  },
  // Links here the same definition given in the Entity Type meta JSON Schema.
  "links": {
    "type": "object",
    "propertyNames": {
      "$comment": "Property names must be a valid URI to a link-type",
      "type": "string",
      "format": "uri"
    },
    "patternProperties": {
      ".*": {
        "type": "object",
        "oneOf": [
          {
            "properties": {
              "ordered": { "type": "boolean", "default": false },
              "type": { "const": "array" }
            },
            "required": ["ordered", "type"]
          },
          {}
        ],
        "additionalProperties": false
      }
    }
  }
}
```

The `kind` can as explained in the Guide-Level explanation only take on values that specify the types of the proposed type system.

The `links` keyword specifically allows constraints in the case of having a set of links. Here the link set can be ordered or unordered (default).

As mentioned in the Guide-Level explanation, `description` for Link Types add semantic meaning to a link, which could be encapsulated in a vocabulary specification for the meta schemas.

> For the most part, we're using existing JSON Schema keywords, but it would be preferable to look into defining these additions as Vocabularies or otherwise integrate them into a JSON Schema validator used for the proposed system.

Reviewer 2: Thanks, addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/a7f4a1fc8a96f2679c843a79c5be6907f7a62886 and is more explicit about changes

## Using the Types in the Block Protocol

Using the proposed type system for Block Protocol imposes changes on the Graph Service and how Block Schemas are defined.

> The Graph Service is currently specified as [this schema](https://github.com/blockprotocol/blockprotocol/blob/main/packages/%40blockprotocol/graph/src/graph-service.json).

Author: Thank you!

**An example of an [`Entity`](https://blockprotocol.org/types/services/graph/entity) instance in the current system:**

```json
{
  "entityId": 111,
  "properties": {
    "name": "Arthur Philip Dent",
    "age": 30,
    "planetOfOrigin": "Earth",
    "occupation": "Intergalactic Traveler"
  }
}
```

**An example of an entity instance in the proposed system:**

```json
{
  "https://blockprotocol.org/types/@blockprotocol/property-type/entity-id": 111,
  "https://blockprotocol.org/types/@alice/property-type/name": "Arthur Philip Dent",
  "https://blockprotocol.org/types/@alice/property-type/age": 30,
  "https://blockprotocol.org/types/@alice/property-type/planet-of-origin": "Earth",
  "https://blockprotocol.org/types/@alice/property-type/occupation": "Intergalactic Traveler"
}
```

This change canonicalizes property keys, such that they uniquely identify a Property Type.

**Within blocks** this impacts:

- how entity fields are accessed
- how Block Protocol messages are constructed
- payload size of Block Protocol messages

**For block authors** this impacts:

- how Block Schemas are defined

> - reusability of properties (the need for some kind of type discovery)

Reviewer 4: Can we unpack this? To the extent that BP users wanted to use other people's types in the current system, there would already be a need for some kind of type discovery, and this proposal doesn't increase the need to use other people's types (I suppose arguably out of convenience since it's now more work to define them?)

Reviewer 4: ```suggestion
- the reusability of properties and data types (and the increased need for type discovery to facilitate this)
```

> 💡 Notice that `entityId`s are replaced by a Property Type URI that resides in the `/@blockprotocol` namespace. This is an implementation detail, and not something that is strictly dictated by the proposal.

### Interfacing with Linked Entities

#### Receiving Links and Linked Entities

Link groups and Linked entities in the Graph Service are currently supplied outside the entity as separate objects. This behavior will stay the same, but the objects received will be of a different shape.

**An example of a [`LinkGroup`](https://blockprotocol.org/types/services/graph/link-group) instance in the current system:**

```json
{
  "sourceEntityId": "user1",
  "path": "company",
  "links": [
    {
      "sourceEntityId": "user1",
      "destinationEntityId": "company1",
      "path": "company"
    }
  ]
}
```

**An example of a `LinkGroup` instance in the proposed system:**

```json
{

>   "sourceEntityId": 111,

Reviewer 4: Is there a reason the ids have changed? If not, I'd keep them consistent so it's easier to see what the proposal actually changes

Author: Just an attempt to keep things simple, seems like it's proven distracting instead so we should update 👍

Reviewer 2: Addresses here https://github.com/blockprotocol/blockprotocol/pull/352/commits/810a0fd6ca7c82655da3809bb9d5575fda297681 Thanks!

  "schema": {
    "properties": {
      "birthDate": { "type": "string", "format": "date" }
    }
  }
}
```

**Entity Type Create and Update messages in the proposed system:**

```json
{
  // createEntityType message
  "schema": {
    "type": "object",
    "kind": "entityType",
    "name": "Person",
    "properties": {
      "https://blockprotocol.org/types/@alice/property-type/name": {
        "$ref": "https://blockprotocol.org/types/@alice/property-type/name"
      },
      "https://blockprotocol.org/types/@alice/property-type/email": {
        "$ref": "https://blockprotocol.org/types/@alice/property-type/email"
      },
      "https://blockprotocol.org/types/@alice/property-type/phone-number": {
        "$ref": "https://blockprotocol.org/types/@alice/property-type/phone-number"
      }
    }
  }
}
```

```json
{
  // updateEntityType message
  "entityTypeId": "Person",
  "schema": {
    // The properties here are partially applied to the original Entity Type.
    "properties": {
      "https://blockprotocol.org/types/@alice/property-type/birth-date": {
        "$ref": "https://blockprotocol.org/types/@alice/property-type/birth-date"
      }
    }
  }
}
```

> The update messages of both the current and new systems make use of partial schemas, merging the schema given in the message contents with the existing Entity Type. This may or may not be the desired semantics of updating, and could lead to undesired behavior. In that case, the semantics can be changed to treat updates as a complete replacement.

Reviewer 4: By whom, how? Do we mean on a per-operation basis, or that we might change the proposal?

Reviewer 2: I think the original text is a little misleading. The semantics around updating is sort of parallel to inheritance/forking, so while this proposal might not change much, the RFC that will address inheritance/forking will bring in the considerations around updating types.

> 💡 The `entityTypeId` here refers to the unique name of the Entity Type. It could make use of the Entity Type's canonical URI instead.

**Property Type Create and Update messages in the current system:**

Property Types do not exist.

**Property Type Create and Update messages in the proposed system:**

```json
{
  // createPropertyType message
  "schema": {
    "kind": "propertyType",
    "name": "User ID",
    "oneOf": [
      {
        "$ref": "https://blockprotocol.org/types/@blockprotocol/data-type/text"
      }
    ]
  }
}
```

```json
{
  // updatePropertyType message

>   "propertyTypeId": "User ID",

Reviewer 4: `name` is not a unique identifier, surely this should be whatever its `$id` is?

Reviewer 2: Addressed in https://github.com/blockprotocol/blockprotocol/pull/352/commits/644bfc128e2b0c55e907edff429e6dfa3843dab3, fixing other similar places

## Increased dependency on external parties

A potential negative side-effect of building an open ecosystem is introducing more dependencies on external parties. Although the magnitude of the impact will depend on the accompanying tooling and implementation that comes about as a result of this RFC, generally speaking there will be an increased reliance on external parties hosting and maintaining types. That said, the Block Protocol is focused around front-end components, and generally speaking there is already a high-reliance on external ecosystems (whether it be npm, etc.), so this is a risk that should be familiar for developers. Some of the [Future possibilities](#future-possibilities) outlined later on would also help mitigate this with potential centralized repositories for types to improve confidence in guarantees and consistency, etc.

## Ergonomics of manually writing schemas

> This RFC greatly decreases the ergonomics of writing block schemas **by hand**. The verbosity of schemas is significantly increased, as well as the general complexity of their expression due to them being composed of other schemas that need to be referred to.

Reviewer 4: ```suggestion
This RFC greatly decreases the ergonomics of writing block schemas **by hand**. The verbosity of schemas is significantly increased, as well as the general complexity of their expression due to them being composed of other schemas that need to be referred to, and which need to be addressable at some unique location at the point of production use.
```

Author: Also good to highlight, thank you

This has the potential to increase the barrier-to-entry, and dissuade people from picking up the Protocol if its capabilities prove to be too complicated for their use cases.

## Ecosystem lock-in

With some of the potential solutions proposed in [Future Possibilities](#future-possibilities) there will be large benefits to participating within the Block Protocol ecosystem and community. This naturally brings the potential downside of the opportunity cost of _not_ participating within the ecosystem.

This type system is designed around reusable types, where people can share and benefit from other people's work, isolating development from the community means that types will not end up being reusable and a lot of the value is lost.

## Versioning Considerations

> A version control model will likely be developed in tandem to the implementation of this type system. As such, there are potential complications to how dependencies are managed and the ramifications of having a large-connected system. Some of these are explored in detail in the [RFC about constraining Links](https://github.com/blockprotocol/blockprotocol/pull/354). The major drawback here is that this greatly increases cognitive load, and reasoning about changes within the ecosystem will become more involved.

Reviewer 4: ```suggestion
A version control model will likely be developed in tandem to the implementation of this type system. As such, there are potential complications to how dependencies are managed, how types are addressed, and the ramifications of having a large-connected system. Some of these are explored in detail in the [RFC about constraining Links](https://github.com/blockprotocol/blockprotocol/pull/354). The major drawback here is that this greatly increases cognitive load, and reasoning about changes within the ecosystem will become more involved.
```

### Allowing `oneOf`

Representing the Type System within common structures like a Relation Database is greatly complicated by the inclusion of the boolean operators (`oneOf` is equivalent to an `or` operator). It is required under the constraint of being able to map any existing JSON data, as reasoned above. A simple example of this is trying to represent an object (this case occurs either in an Entity or a Property Type object) that looks like the following:

```json
{
  "someProperty": [2.0, "foo"]
}
```

To be able to represent this object, a Property Type would need to be created that looks like (a simplified view for brevity):

```json
{
  "name": "someProperty",
  "oneOf": [{ "$ref": "data-type/number" }, { "$ref": "data-type/text" }]
}
```

More specifically, it's required to allow `oneOf` with more variants than _just Data Types_ because:

```json
{
  "someProperty": [{ "bar": 2.0 }, "foo"]
}
```

Requires something like

```json
{
  "name": "someProperty",
  "oneOf": [
    {
      "type": "object",
      "properties": {
        "bar": { "$ref": "property-type/bar" }
      }
    },
    { "$ref": "data-type/text" }
  ]
}
```

### Including an `Empty List` Data Type

> Similarly, the `Empty List` Data Type is required to be able to programmatically map unknown JSON data into the system. Given an incoming entity like:

Reviewer 4: I think this is what we mean, since it is possible to map this as an `array` with no constraints on its `items` without introducing an `Empty List`. 
```suggestion
Similarly, the `Empty List` Data Type is required to be able to programmatically map unknown JSON data into the system without typing it as an unconstrained list. Given an incoming entity like:
```

Author: That's right. The general idea is that we don't want a plain "list" type because Update methods and such that could rely on types for validation might allow a user to expand the list and add anything they want, implicitly (and undesirably) adding some notion of an "any" type. By not allowing the list to have more elements added to it, it will be required to change the type to explicitly define what is inside the list.

# Prior art

[prior-art]: #prior-art

As mentioned in a few sections, this design basically defines a way for communities to build a compound ontology. As such it's worth mentioning some of the alternative knowledge representation models to provide points of discussion, and to invite members of their communities to give thoughts. There are a few specific technologies that we've taken particular interest in while designing the RFC.

> Note: Some planning has been made around **varying**-levels of compatibility with these technologies, although in-depth designs for their interfacing have not been written up yet.

## Data Description Formats

- [RDF](https://www.w3.org/RDF/)
- [RDFS](https://www.w3.org/TR/rdf-schema/)
- [OWL](https://www.w3.org/OWL/)

## Ontologies

- [Schema.org](https://schema.org/)
- [DBPedia](https://dbpedia.org/)

# Unresolved questions

[unresolved-questions]: #unresolved-questions

1.  Are there further impacts on block schemas?

> 1.  Do we permit spaces in type names?

Reviewer 4: Why is this a question? Aren't they just display names?

Reviewer: > I'm inclined to think we should treat them as plain text and spaces are fine.

That's also my instinct

> perhaps there are cases where EAs would benefit from these not having spaces?

Perhaps, but I can't think of a reason - perhaps we can get rid of this unresolved question as we are all agreeing we like spaces in display names, and can't come up with a reason why spaces are problematic?

Reviewer 2: Addressed the immediate decision in https://github.com/blockprotocol/blockprotocol/pull/352/commits/0bd6547994b1a205523473d5230633554d181c9a

> 1.  Is there a way to specify in JSON schema that the key of a property is equal to the the thing it's a `$ref` to?

Reviewer 4: It would help to have an explanation somewhere that we want this / why is it desirable

Reviewer 2: I've added some technical reason why (https://github.com/blockprotocol/blockprotocol/pull/352/commits/7842a041d3cdd435776db3194e097ee1c4073d86), we might also have some spec-reason why this is desirable

Reviewer 4: Thank you – this addresses why we want to be able to _validate_ this, although we still don't explain anywhere why we want them to be the same at all.

We can track the need for that in this [thread](https://github.com/blockprotocol/blockprotocol/pull/352#discussion_r910774375), so I'll resolve this.

    As in can we specify a constraint that you have to have equal URIs in `"someUri": { "$ref": "someUri" }`?
1.  Should we further constrain allowed URIs (for example to force the end of the path to be `/property-type/foo`) and if so should we encode that in the JSON schema?

> 1.  Do we want to allow types to define a plural name that can be used when they're set to `"type": "array"`?

Reviewer 4: ```suggestion
1.  Do we want to allow types to define a separate plural name that can be used when they're set to `"type": "array"`?
```

The `kind` keyword will be used to identify the kind of the schema being described. The value of `kind` must be one of:

- `entityType`
- `propertyType`
- `dataType`
- `linkType`

The `links` keyword will be used to allow Entity Type schemas to define links. The value of `links` is an object, whose keys are URIs (that point to Link Types). Values of `links` are also objects, which can optionally define extra constraints on the link (for more information see the [Reference-Level explanation](#reference-level-explanation)).

The `requiredLinks` keyword will be used to specify which links are required to be present for a given Entity Type. It is a list of URIs.

In JSON Schema `title` and `description` are described as follows: "Both of these keywords can be used to decorate a user interface with information about the data produced by this user interface".
We repurpose these with slightly different meanings:

The `description` keyword for Link Types adds semantic meaning to a link (which could be encapsulated in a custom vocabulary specification for the meta schemas).

The `title` keyword refers to the name of the Type it is describing.

> The `$ref` keyword when used in the context of Entity Types, Property Types, Data Types, and Link Types URIs to specify references across the different kinds of types, the `$ref` will always need to be equal the key of the object which it is defined under. If the `$ref` is within an array definition, the URI should equal the nearest JSON Schema property name which is a URI.

Reviewer 4: The other way of thinking of this is that the constraint is on _the key_, rather than on the `$ref`. It's not that users pick a key and then make the `$ref` the same – they choose the `$ref` and then set the key accordingly.

I haven't captured that here as the effect is the same, but worth thinking about in the context of validation approaches. I do think we could explain _why_ we want this restriction, and have made a suggestion to that effect.

```suggestion
The `$ref` keyword when used to refer to Property Types, Data Types, and Link Types URIs will always need to be equal the key of the object which it is defined under (and vice versa). If the `$ref` is within an array's `items` definition, the URI should equal the nearest JSON Schema property name which is a URI. This is so that the property, data, or link type can be identified from an instance conforming to it (which it could not if the key, i.e. the property name, were different from the `$ref`).
```

These elements are used in combination to improve reusability of, and better define, expressions of data structural requirements.

> By using the proposed types, blocks, embedding applications, and end-users can all have greater confidence in each other's roles and functionalities, and hopefully will allow them all to benefit from better computational inference and capabilities. Reusability of descriptions of data should decrease developmental overhead, with all members of the ecosystem benefiting from emergent standardization.

Author: I appreciate this was quite wordy but

> should encourage a decrease in developmental overhead over time

Is more correct. The sentence was highlighting that developmental overhead will decrease _over time_. For a long time it will actually increase, and this is a risk we have to be aware of. It's only when we reach some form of critical mass that people will start to largely benefit from it.

By getting both parties to agree to a prescriptive standard, new technological possibilities become viable. The Block Protocol is an example of this, enabling a new class of applications, ones that are able to dynamically load complex front-end components at run-time which are immediately usable and functional, even if the component was previously unknown to the application.

> The foundation of this capability is standardizing how requirements are expressed. Most notably, the block and entity schemas expressed through the graph service allow for dynamic resolution of expected input data.

Author: I think we should avoid changes of grammar like this because it feels like wasted time when it doesn't affect the content of the document. However, because it's been made, I don't think this phrasing is as grammatically correct. It sounds like the foundation is the thing that standardizes, rather than the standardization of X _is_ the foundation.

Author: @benwerner01 in case this has gotten lost

    },
    "title": { "type": "string" },
    "description": { "type": "string" },
    "default": {
      "$comment": "Default Entity instance",
      "type": "object",
      "propertyNames": {
        "$comment": "Property names must be a valid URI to a Property Type",
        "type": "string",
        "format": "uri"
      }
    },

>     "example": {

Reviewer 4: Also, definitely not for now, but we should define things like `propertyNames` once in `$defs` instead of repeatedly throughout the schema.
```suggestion
    "examples": {
```
